### DISCUSSION TITLE
esp32-3248s035c light sensor onboard

### DISCUSSION THREAD

**Discussion by samoswall** [HASwitchPlate/openHASP#393] [Q&A]
Please tell me how to add a built-in light sensor on the board esp32-3248s035c.
![openHASP plate light sensor](https://user-images.githubusercontent.com/80251553/196060819-d69f57d5-ccd2-40a4-ae35-39bad1e458ea.png)
The light sensor is connected to GPIO34 (ADC6)
![ESP32-3248S035-ADC](https://user-images.githubusercontent.com/80251553/196060988-4ae00f28-7b58-447c-b5d6-46e7b32eeea1.jpg)
As I understand it, you need to add a Custom Code.
In GPIO Settings adc is missing.
I know that many have already purchased the esp32-3248s035, maybe it's worth adding ADC light sensor to the configuration esp32-3248s035 board.

**Comment by JohnJJG** [comment]
I just got my esp32-3248s035C yesterday so wanted to do this as well.  The code below will send light readings and allow the led's to be set

```
// USAGE: - Copy this file to src\custom\my_custom_esp32_3248s035C.cpp, also copy my_custom_template.h and rename it to my_custom.h
//        - Uncomment in your user_config_override.h the line containing #define HASP_USE_CUSTOM 1
//        - Led state can be set with hasp/<yourplate>/custom/led and payload {"red":1,"green":1,"blue":1} 
//          Leds are on off only so use 1 or 0 on each led
//        - Light level will be sent with the topic hasp/<yourplate>/custom/light
//          Light level will update every 5 second only if it has changed by more than the tolerance value
#include "hasplib.h"

#if defined(HASP_USE_CUSTOM) && true

#include "hasp_debug.h"

#define LIGHT_SENSOR 34
#define LED_RED 4
#define LED_GREEN 16
#define LED_BLUE 17

int iLastLightLevel = 0; // light reading from the sensor
unsigned int iLightTolerance = 10; // to reduce messages only send light message if change is outside tolerance

void custom_setup()
{
    // intialise light sensor
    pinMode(LIGHT_SENSOR, ANALOG);
    analogSetPinAttenuation(LIGHT_SENSOR, ADC_0db); // 0dB(1.0) 0~800mV
    iLastLightLevel = analogReadMilliVolts(LIGHT_SENSOR);
    
    // inialise led (off)
    pinMode(LED_BLUE, OUTPUT);
    pinMode(LED_RED, OUTPUT);
    pinMode(LED_GREEN, OUTPUT);

    digitalWrite(LED_BLUE, HIGH); // off is HIGH, on is LOW 
    digitalWrite(LED_RED, HIGH); // off is HIGH, on is LOW 
    digitalWrite(LED_GREEN, HIGH); // off is HIGH, on is LOW 
}

void custom_loop()
{
}

void custom_every_second()
{
}

void custom_every_5seconds()
{
    char sLightLevel[32];

    // send light level reading if changed
    int iNewLightLevel = analogReadMilliVolts(LIGHT_SENSOR);
    if ((iNewLightLevel > (iLastLightLevel + iLightTolerance)) || (iNewLightLevel < (iLastLightLevel - iLightTolerance))) {
        snprintf_P(sLightLevel, sizeof(sLightLevel), PSTR("{\"light\":%d}"), iNewLightLevel);
        dispatch_state_subtopic("light",sLightLevel);
        iLastLightLevel = iNewLightLevel;
    }
}

bool custom_pin_in_use(uint8_t pin)
{
    switch(pin) {
        case LIGHT_SENSOR:
        case LED_BLUE:
        case LED_RED:
        case LED_GREEN:
            return true;
        default:
            return false;
    }
}

void custom_get_sensors(JsonDocument& doc)
{
}

void custom_topic_payload(const char* topic, const char* payload, uint8_t source){
    LOG_INFO(TAG_CUSTOM, "Handling esp32-3248s035C custom message: %s => %s", topic, payload);
    
    // set the led bit1: red, bit2: green, bit3: blue
    if(!strcmp(topic, "led")) {
        StaticJsonDocument<1024> led_doc;
        deserializeJson(led_doc, payload);

        digitalWrite(LED_RED, led_doc["red"] == 1 ? LOW : HIGH); 
        digitalWrite(LED_GREEN, led_doc["green"] == 1  ? LOW : HIGH); 
        digitalWrite(LED_BLUE, led_doc["blue"] == 1  ? LOW : HIGH); 
    }
}

#endif // HASP_USE_CUSTOM
```

**Comment by JohnJJG** [comment]
If you would prefer the light level to be in the sensor block then I have modified the code above with the code moved from custom_every_5seconds() into custom_get_sensors().  I am getting about 75 if there is light and 1080 if it is completely dark

![image](https://user-images.githubusercontent.com/4007912/204958345-d14a60c3-d8d5-4e55-88c2-fd6cab201849.png)

```
void custom_every_5seconds()
{
}

void custom_get_sensors(JsonDocument& doc)
{
    // send light level reading if changed
    int iNewLightLevel = analogReadMilliVolts(LIGHT_SENSOR);
    if ((iNewLightLevel > (iLastLightLevel + iLightTolerance)) || (iNewLightLevel < (iLastLightLevel - iLightTolerance))) {
        iLastLightLevel = iNewLightLevel;
    }
    doc[F("light")] = iLastLightLevel;  // Set Value
}
```

**Comment by FredericG-BE** [comment]
I have just tried this custom code but iNewLightLevel is always 75. Would it be possible the schematics have changed I wonder... 
I failed to find schematics for the board, here could I find those?
Thanks

**Reply by JohnJJG** [reply]
Hi Frederic,
Have you tried when it is really dark.  The sensor is not very good, I think it will read 75 if it is 1cm from the sun down to near dark.  only on very low light levels to you see any granularity.  Needs to be very dark to get it to register above 75.  It is more a darkness sensor.  With a low light lamp on I get about 100, put it in a drawer I get about 700+.

**Reply by JohnJJG** [reply]
Hi Frederic,
These are the links I downloaded stuff from. The full pack is on the second link (I would scan the file if you download it).  The first one is the summary/review but in Japanese.  I am sure there was an English version but I cannot find it

Review /Summary (Japanese) but there is an english version out there
https://macsbug.wordpress.com/2022/10/02/esp32-3248s035/

Spec/Schematic pack (from the AliExpress Sunton Store)
http://pan.jczn1688.com/directlink/1/ESP32%20module/3.5inch_ESP32-3248S035.zip?spm=a2g0o.detail.1000023.1.2e10zRLDzRLDjm&file=3.5inch_ESP32-3248S035.zip

**Reply by FredericG-BE** [reply]
Thanks John for de details. 
In my case, when I cover the sensor or shine in it with a lamp, the value stays put... 
Putting IO34 on a scope shows that the voltage swings due to different lighting conditions are less than the noise on the signal.  I wonder how this is supposed to work; the variation on the LDR are much too small for the 1MOhm R15. I soldered a 10K over R15 and I see now variations between 300 and 1000 depending on the light. Did they start using another LDR without changing R15 I wonder...
Anyway, seems to work now... thanks
